(CDAP-7670) Implemented backend for collecting and reporting operatio…

bdmogal opened the pull request:
…nal stats using JMX

- Added a new cdap-operational-stats-core module
- Classes from cdap-operational-stats-core that implement OperationalStats will be loaded as extensions using the Java ServiceLoader architecture, and be registered as JMX Beans
- As initial integration, added operational stats for HDFS
- Also added REST APIs and JMX Reader for reporting JMX stats

Jira: [CDAP-7670](https://issues.cask.co/browse/CDAP-7670)
Build: http://builds.cask.co/browse/CDAP-DUT5085

There are no tests in this PR yet, but wanted to get an initial review done first, since the PR was becoming large. Tested this to work on a cluster for HDFS.

Commit:
(CDAP-7670) Implemented backend for collecting and reporting operational stats using JMX

- Added a new cdap-operational-stats-core module
- Classes from cdap-operational-stats-core that implement OperationalStats will be loaded as extensions using the Java ServiceLoader architecture, and be registered as JMX Beans
- As initial integration, added operational stats for HDFS
- Also added REST APIs and JMX Reader for reporting JMX stats

anwar6953 commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/OperationalStatsHttpHandler.java:
```diff
@@ -0,0 +1,70 @@
+    if (stats.isEmpty()) {
+      throw new NotFoundException(String.format("Service provider %s not found", serviceProvider));
+    }
+    Map<String, Map<String, Object>> filtered = Maps.filterKeys(stats, new Predicate<String>() {
```
this seems like fewer lines of code and more efficient:

``` java
  if (map.containsKey("info")) {
      map.remove("info");
    }
```

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/operations/OperationalStatsCollector.java:
```diff
@@ -0,0 +1,172 @@
+ * class that implements an {@link MXBean} should also additionally extend {@link OperationalStats}. This class loads
+ * implementations of {@link OperationalStats} using the Java {@link ServiceLoader} architecture.
+ */
+@Singleton
```
Better use the singleton at binding time then a `@Singleton` annotation, unless this class won't work if it is not a singleton.

bdmogal replied:
Fixed

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/operations/OperationalStatsCollector.java:
```diff
@@ -0,0 +1,172 @@
+ */
+@Singleton
+public class OperationalStatsCollector
+  extends AbstractExtensionLoader<OperationalExtensionId, OperationalStats> implements Closeable {
```
Seems like better to separate this class into two. A `OperationStatsExtensionLoader` and a `OperationStatsService`, which can have start(), stop() methods.

bdmogal replied:
Fixed

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/operations/OperationalStatsReader.java:
```diff
@@ -0,0 +1,109 @@
+   */
+  public Map<String, Map<String, String>> getStatsOfType(String statType)
+    throws MalformedObjectNameException, IntrospectionException, InstanceNotFoundException, ReflectionException,
+    AttributeNotFoundException, MBeanException {
```
Really need to list them all out? Can't just `throws Exception`?

bdmogal replied:
Fixed

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/operations/OperationalStatsReader.java:
```diff
@@ -0,0 +1,109 @@
+   * @param serviceName the service for which operational stats are required
+   * @return a map of stat type (e.g. storage, compute, etc) to a map of stats of the specific type
+   */
+  public Map<String, Map<String, Object>> getOperationalStats(String serviceName)
```
Seems like this and previous are quite the same, the only different is the call to `getKeyProperty`?

bdmogal replied:
yes, refactored them into a single method.

chtyim commented on cdap-common/src/main/resources/cdap-default.xml:
```diff
@@ -1341,6 +1341,32 @@
+  </property>
+
+  <property>
+    <name>operational.stats.collection.threads</name>
```
Why we need that many threads. From the logic, it seems like only one thread is needed (you use `scheduleWithFixedDelay`, hence it won't have overlapping run).

bdmogal replied:
Although I didn't implement that, my initial thought was to collect a single operational stat per thread. I didn't implement that because I felt it would be unnecessary to start with. Forgot to remove this. Removed now. Will think about multiple threads if there really is a use-case for it.

chtyim reviewed (commented):
Did a round of quick pass

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/operations/OperationalStatsCollector.java:
```diff
@@ -0,0 +1,172 @@
+ */
+@Singleton
+public class OperationalStatsCollector
+  extends AbstractExtensionLoader<OperationalExtensionId, OperationalStats> implements Closeable {
```
Fixed.

bdmogal commented:
@chtyim addressed comments. Please review again.

sagarkapare commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/OperationalStatsHttpHandler.java:
```diff
@@ -0,0 +1,74 @@
+@Path(Constants.Gateway.API_VERSION_3 + "/system/serviceproviders")
+public class OperationalStatsHttpHandler extends AbstractHttpHandler {
+
+  private final OperationalStatsReader operationalExtensionsReader;
```
nit: rename variable to `operationalStatsReader`

sagarkapare commented on cdap-app-fabric/src/main/java/co/cask/cdap/operations/OperationalStatsReader.java:
```diff
@@ -0,0 +1,72 @@
+import javax.management.ObjectName;
+
+/**
+ * Reads operational stats collected using {@link OperationalStatsReader}.
```
nit: please add more relevant javadoc.

bdmogal replied:
fixed.

sagarkapare commented on cdap-app-fabric/src/main/java/co/cask/cdap/operations/OperationalStatsReader.java:
```diff
@@ -0,0 +1,72 @@
+   * @return a {@link Map} of the group to a {@link Map} of stats of that group
+   */
+  public Map<String, Map<String, Object>> getStats(String propertyKey, String propertyValue,
+                                                   String groupByKey) throws Exception {
```
would be good list the exceptions thrown by this method in the javadoc.

sagarkapare replied:
should we make this method static?

bdmogal replied:
this method throws a whole bunch of exceptions thrown by the JMX classes. I'm not sure it will be helpful to users of this class. Maybe I can describe *when* exceptions will be thrown.

sagarkapare commented on cdap-app-fabric/src/main/java/co/cask/cdap/operations/OperationalStatsReader.java:
```diff
@@ -0,0 +1,72 @@
+/**
+ * Reads operational stats collected using {@link OperationalStatsReader}.
+ */
+public class OperationalStatsReader {
```
better make this class `final`

sagarkapare commented on cdap-app-fabric/src/main/java/co/cask/cdap/operations/OperationalStats.java:
```diff
@@ -0,0 +1,47 @@
+import javax.management.MXBean;
+
+/**
+ * Base class for all operational stats emitted using the operational stats extension framework.
```
nit: interface

bdmogal replied:
fixed

sagarkapare commented on cdap-operational-stats-core/src/main/java/co/cask/cdap/operations/hdfs/HDFSNodesMXBean.java:
```diff
@@ -0,0 +1,29 @@
+  /**
+   * Returns number of namenodes
+   */
+  int getNamenodes();
```
should we have `getDataNodes()` as well or not required?

bdmogal replied:
It is not possible to get that info right now, since the API to get it expects that the caller is the "hdfs" user. I will create a jira to figure out a way of retrieving that information.

bdmogal replied:
https://issues.cask.co/browse/CDAP-7679

sagarkapare commented on cdap-operational-stats-core/src/main/java/co/cask/cdap/operations/hdfs/HDFSInfo.java:
```diff
@@ -0,0 +1,121 @@
+  private String getNameService() {
+    Collection<String> nameservices = conf.getTrimmedStringCollection(DFSConfigKeys.DFS_NAMESERVICES);
+    if (nameservices.isEmpty()) {
+      return null;
```
When would `nameservices` be empty? If its only in case if federated NS is configured, should we throw `IllegalStateException` here?

bdmogal replied:
It can be null when there is no federated HDFS, which also means no HA. If it is present, it should never be empty, because ``getTrimmedStringCollection`` will return a collection with at least one element. Fixed.

However, in the first case (when nameservices is null), it is better to return null, because some users of this API (e.g. ``HAUtil.isHAEnabled()``, expect this to return null when federation is not enabled).

bdmogal replied:
actually, I was wrong: ``getTrimmedStringCollection`` can return an empty collection when the property is not configured. However, in that case, this method needs to return a null, because that's the contract of other methods that use this - e.g. ``HAUtil.isHAEnabled()``. I've added a comment.

sagarkapare commented on cdap-gateway/src/test/java/co/cask/cdap/gateway/router/RouterPathTest.java:
```diff
@@ -433,6 +433,24 @@ public void testPreviewPaths() {
+  @Test
+  public void testServiceProviderStatsPaths() {
+    String serviceProviderPath = "/v3/system/////serviceproviders";
+    HttpRequest httpRequest = new DefaultHttpRequest(VERSION, new HttpMethod("GET"), serviceProviderPath);
```
Does the http method matters? If not we can simply use `assertRouting` method.

bdmogal replied:
thanks! fixed.

sagarkapare commented on cdap-common/src/main/resources/cdap-default.xml:
```diff
@@ -1341,6 +1341,24 @@
+    <name>operational.stats.extensions.dir</name>
+    <value>/opt/cdap/master/ext/operations</value>
+    <description>
+      Semicolon-separated list of local directories that are scanned for
```
local directories to cdap-master?

bdmogal replied:
fixed

sagarkapare reviewed (commented):
took one pass. mostly looks good. minor comments.

sagarkapare commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/OperationalStatsHttpHandler.java:
```diff
@@ -0,0 +1,65 @@
+/**
+ * {@link co.cask.http.HttpHandler} for service provider statistics.
+ */
+@Path(Constants.Gateway.API_VERSION_3 + "/system/serviceproviders")
```
should be `@Singleton`

bdmogal replied:
Not all handlers are Singleton. What's our policy on it? This one in particular doesn't have any state, so it should be ok not to mark it?

sagarkapare commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/OperationalStatsHttpHandler.java:
```diff
@@ -0,0 +1,65 @@
+    responder.sendJson(
+      HttpResponseStatus.OK,
+      OperationalStatsUtils.getStats(OperationalStatsUtils.STAT_TYPE_KEY,
+                                      OperationalStatsUtils.STAT_TYPE_INFO,
```
nit: indentation

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/operations/OperationalStatsUtils.java:
```diff
@@ -0,0 +1,105 @@
+   * @return a {@link Map} of the group to a {@link Map} of stats of that group
+   * @throws Exception when there are errors reading stats using JMX
+   */
+  public static Map<String, Map<String, Object>> getStats(String propertyKey, String propertyValue,
```
Better move this as private in the handler class.

sagarkapare commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/OperationalStatsHttpHandler.java:
```diff
@@ -0,0 +1,113 @@
+   */
+  @VisibleForTesting
+  Map<String, Map<String, Object>> getStats(String propertyKey, String propertyValue,
+                                                          String groupByKey) throws Exception {
```
nit: indentation

sagarkapare reviewed (commented):
minor comments, rest LGTM

bdmogal commented:
@chtyim @sagarkapare addressed comments. The last two commits also contain a couple of unit tests, which were missing earlier since this PR was getting big. Please review.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/operations/OperationalStatsService.java:
```diff
@@ -0,0 +1,160 @@
+        entry.getValue().collect();
+      } catch (IOException e) {
+        LOG.warn("Exception while collecting stats for service: {}; type: {}", entry.getValue().getServiceName(),
+                 entry.getValue().getStatType());
```
log the exception `e` as well.

chtyim commented on cdap-operational-stats-core/src/main/java/co/cask/cdap/operations/hdfs/AbstractHDFSStats.java:
```diff
@@ -0,0 +1,52 @@
+
+  protected DistributedFileSystem createDFS() throws IOException {
+    FileSystem fs = FileSystem.get(conf);
+    Preconditions.checkArgument(fs instanceof DistributedFileSystem, "Expected Distributed Filesystem to be the " +
```
So what happen if this exception is throw? Just that this extension will get skipped?

bdmogal replied:
yup. it will get skipped.

chtyim commented:
one comment and one question, the rest is good.

bdmogal commented:
Addressed final comment. Waiting for http://builds.cask.co/browse/CDAP-DUT5085-9

chtyim commented:
Merge conflict. I believe you have to resolve it before running the build.

bdmogal commented:
Hmmm, the build had started a while ago. I guess another PR got merged after that. Will rebase

bdmogal commented:
Build succeeded: https://github.com/caskdata/cdap/pull/7142. Last couple of commits were only merge conflict resolutions, and doc changes. Squashed commits. Merging.
